File: simpletranscribe-mac/simpletranscribe/Services/TranscriptExporter.swift
import Foundation
import UniformTypeIdentifiers
import AppKit

// MARK: - Format

enum ExportFormat: String, CaseIterable, Identifiable {
    case txt, md, srt
    var id: String { rawValue }

    var displayName: String {
        switch self {
        case .txt: return "Plain Text (.txt)"
        case .md:  return "Markdown (.md)"
        case .srt: return "SubRip (.srt)"
        }
    }

    var contentType: UTType {
        switch self {
        case .txt: return .plainText
        case .md:  return UTType("net.daringfireball.markdown") ?? .plainText
        case .srt: return UTType("com.scenarist.closed-caption-srt") ?? .plainText
        }
    }
}

// MARK: - Exporter

enum TranscriptExporter {

    static func formatText(_ text: String) -> String { text }

    static func formatMarkdown(_ text: String, title: String = "Transcript") -> String {
        "# \(title)\n\n\(text)\n"
    }

    static func formatSRT(entries: [TranscriptEntry]) -> String {
        guard !entries.isEmpty else { return "" }
        var offset: TimeInterval = 0
        return entries.enumerated().map { (i, entry) in
            let start = srtTime(offset)
            let end   = srtTime(offset + max(entry.duration, 1))
            offset   += entry.duration
            return "\(i + 1)\n\(start) --> \(end)\n\(entry.text)\n"
        }.joined(separator: "\n")
    }

    // MARK: - Helpers

    private static func srtTime(_ seconds: TimeInterval) -> String {
        let h  = Int(seconds / 3600)
        let m  = Int(seconds.truncatingRemainder(dividingBy: 3600) / 60)
        let s  = Int(seconds.truncatingRemainder(dividingBy: 60))
        let ms = Int((seconds - floor(seconds)) * 1000)
        return String(format: "%02d:%02d:%02d,%03d", h, m, s, ms)
    }
}

File: simpletranscribe-win/SimpleTranscribe/Services/TranscriptExporter.cs
using SimpleTranscribe.Models;

namespace SimpleTranscribe.Services;

public static class TranscriptExporter
{
    public static string FormatText(string text) => text;

    public static string FormatMarkdown(string text, string title = "Transcript") =>
        $"# {title}\n\n{text}\n";

    public static string FormatSrt(IEnumerable<TranscriptEntry> entries)
    {
        var parts = new List<string>();
        double offset = 0;
        int i = 1;
        foreach (var entry in entries)
        {
            var start = SrtTime(offset);
            var end   = SrtTime(offset + Math.Max(entry.DurationSeconds, 1));
            offset   += entry.DurationSeconds;
            parts.Add($"{i}\n{start} --> {end}\n{entry.Text}\n");
            i++;
        }
        return string.Join("\n", parts);
    }

    public static string FormatHistoryMarkdown(IEnumerable<TranscriptEntry> entries)
    {
        var body = string.Join("\n\n", entries.Select(e =>
            $"## {e.Timestamp:g}\n\n{e.Text}"));
        return FormatMarkdown(body, "Transcript History");
    }

    private static string SrtTime(double seconds)
    {
        var h  = (int)(seconds / 3600);
        var m  = (int)(seconds % 3600 / 60);
        var s  = (int)(seconds % 60);
        var ms = (int)((seconds - Math.Floor(seconds)) * 1000);
        return $"{h:D2}:{m:D2}:{s:D2},{ms:D3}";
    }
}
